src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel_ExpandCollapseItems.cs

```diff
@@ -0,0 +1,36 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Collections;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    internal partial class DocumentOutlineViewModel
+    {
+        /// <summary>
+        /// Queue to expand/collapse the items in the tree view
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<ExpansionOption> _expandCollapseQueue;
+
+        public void EnqueueExpandOrCollapse(ExpansionOption option)
+            => _expandCollapseQueue.AddWork(option, cancelExistingWork: true);
+
+        private async ValueTask ExpandOrCollapseItemsAsync(ImmutableSegmentedList<ExpansionOption> expansionOptions, CancellationToken token)
+        {
+            var expansionOption = expansionOptions.Last();
+
+            // guard here as we are going to be modifying properties on the view models
+            // even though they are bools and this is safe to do without a guard it makes things happen 'in order' for the user
```

Comment:
it's not clear to mea what you're referring to.  Note that if you have multiple concurrent queues, this will certainly ensure that onely one queue at a time can be modifying state.  But it doesn't ensure "in order" if the queues run concurrently.  The only way to get things "in order" is to either have all the mutations done by one queue (since it only executes one thing at a time, and it ensures that the operations passed into the callback are in the same order as they were added to it), or having one queue enqueue into the next queue. 

Anything else would allow the user to make independent operations A and B but have tehm execute as B then A.